Sparse develop

nychiang opened the pull request:
Fix issue #109 and #110.
Note that for ticket #110, I got some different error messages, in which they all complained about some overriding issue.
Tried to add some empty functions in [matrixTestsRajaSparseTriplet.hpp](https://github.com/LLNL/hiop/blob/bbb15a551c6b32e61113baa852df99df82fe2555/tests/LinAlg/matrixTestsRajaSparseTriplet.hpp#L88) and [here](https://github.com/LLNL/hiop/blob/bbb15a551c6b32e61113baa852df99df82fe2555/tests/LinAlg/matrixTestsRajaSparseTriplet.hpp#L93), and now hiop pass the unit tests. 

Has been successful build on Summit, and passed ORNL pipelines.

pelesh commented:
I tried to build `sparse-develop`, but I get a compile error when building `testMatrixSparse.cpp.o` object. See error output below:

```shell
[ 87%] Building CUDA object tests/CMakeFiles/testMatrixSparse.dir/testMatrixSparse.cpp.o
nvcc
(...)
tests/testMatrixSparse.cpp -o testMatrixSparse.cpp.o
/people/pele118/src/pnnl_hiop/hiop/tests/LinAlg/matrixTestsSparse.hpp(765): error: no operator "++" matches these operands
            operand types are: std::initializer_list<int> ++

/people/pele118/src/pnnl_hiop/hiop/tests/LinAlg/matrixTestsSparse.hpp(777): error: no operator "<" matches these operands
            operand types are: std::initializer_list<int> < const hiop::tests::local_ordinal_type

/people/pele118/src/pnnl_hiop/hiop/tests/LinAlg/matrixTestsSparse.hpp(778): error: no operator "+" matches these operands
            operand types are: std::initializer_list<int> + hiop::tests::local_ordinal_type

/people/pele118/src/pnnl_hiop/hiop/tests/LinAlg/matrixTestsSparse.hpp(795): error: no operator "+" matches these operands
            operand types are: hiop::tests::local_ordinal_type + std::initializer_list<int>

/people/pele118/src/pnnl_hiop/hiop/tests/LinAlg/matrixTestsSparse.hpp(796): error: no operator "+" matches these operands
            operand types are: hiop::tests::local_ordinal_type + std::initializer_list<int>
```

pelesh commented:
@ashermancinelli, can you edit CI build scripts to set correct COINHSL dir on newell?

nychiang commented:
In this PR, I and Cosmin tried to build the code with different options and noticed some issues.
In short, I think in cmake, UMPIRE looks for RAJA, and RAJA looks for GPU, i.e., RAJA was built on Summit with GPU support.
Will create a new issue about this.

nychiang commented:
> @ashermancinelli, can you edit CI build scripts to set correct COINHSL dir on newell?

nychiang commented:
@pelesh  what's your options to build it? This is what Asher reported in #110 , but I didn't observe them when compiling the code.

ashermancinelli reviewed (commented):
Great work!

ashermancinelli commented:
> > @ashermancinelli, can you edit CI build scripts to set correct COINHSL dir on newell?

Yes, I will take care of this.

pelesh commented:
> @pelesh what's your options to build it? This is what Asher reported in #110 , but I didn't observe them when compiling the code.

Yes, it is. I updated #110.

pelesh commented:
> In this PR, I and Cosmin tried to build the code with different options and noticed some issues.
> In short, I think in cmake, UMPIRE looks for RAJA, and RAJA looks for GPU, i.e., RAJA was built on Summit with GPU support.
> Will create a new issue about this.

RAJA and Umpire should not require GPU support, but if Umpire is built with CUDA backend, you will need to link to CUDA library. Please do submit an issue.

ashermancinelli commented:
> > @ashermancinelli, can you edit CI build scripts to set correct COINHSL dir on newell?
> 
> Yes, I will take care of this.

I have pushed a fix to branch `sparse-develop-coinhsl-ci-fix`, and will open a PR shortly. Note that this will not pass CI due to the issue Slaven has brought up and is described in #110.

pelesh commented on tests/LinAlg/matrixTestsSparse.hpp:
```diff
@@ -732,6 +732,74 @@ class MatrixTestsSparse : public TestBase
+    const local_ordinal_type* B_jCol = getColumnIndices(&B);
+    const local_ordinal_type B_nnz = B.numberOfNonzeros();
+
+    auto nnz_A_need_to_copy{0};
+    for(auto k=0;k<A_nnz;k++){
+      if(A_iRow[k] >= A_rows_st && A_iRow[k] < A_rows_st + n_rows )
+      {
+        nnz_A_need_to_copy++;
+      }
+      // assume matrix element is ordered by row
```
```suggestion
    local_ordinal_type nnz_A_need_to_copy = 0;
    for(local_ordinal_type k=0; k<A_nnz; ++k){
      if(A_iRow[k] >= A_rows_st && A_iRow[k] < A_rows_st + n_rows )
      {
        nnz_A_need_to_copy++;
      }
      // assume matrix element is ordered by row
```
This should fix #110.

cnpetra commented on src/Drivers/Makefile_ex6:
```diff
@@ -0,0 +1,94 @@
+#########################################################################
```
this is a makefile to build ipopt and ipopt-adapter. Used in development only.

nychiang commented on CMakeLists.txt:
```diff
@@ -53,9 +57,9 @@ if(HIOP_USE_MPI)
 # The binary dir is already a global include directory
-configure_file(
-  "${CMAKE_SOURCE_DIR}/src/Interface/hiop_defs.hpp.in"
-  "${CMAKE_BINARY_DIR}/hiop_defs.hpp")
```
I move these lines to line 236, since I may change the value of option `HIOP_SPARSE`. 
For example, user gives `HIOP_SPARSE=ON`, `HIOP_USE_COINHSL=OFF` and `HIOP_USE_STRUMPACK=OFF`, then there is no valid linear package for sparse linear algebra and I reset option `HIOP_SPARSE=OFF` internally.

nychiang commented on CMakeLists.txt:
```diff
@@ -88,23 +92,23 @@ target_compile_options(hiop_warnings
+  # parent context
   -Wnull-dereference
   -Wdouble-promotion  # Warn on implicit conversion from
-                      # float to double
```
I agree. Not sure why this happens, but I will correct all these issues manually.

nychiang commented on CMakeLists.txt:
```diff
@@ -183,6 +195,48 @@ endif()
 #
 #set(LAPACK_LIBRARIES "-lgfortran;/export/home/petra1/work/installs/xblas-1.0.248/libxblas.a;/home/petra1/work/installs/lapack-3.7.0/libblas.a;/home/petra1/work/installs/lapack-3.7.0/liblapack.a;/home/petra1/work/installs/lapack-3.7.0/liblapack.a;/home/petra1/work/installs/lapack-3.7.0/libblas.a;/export/home/petra1/work/installs/xblas-1.0.248/libxblas.a")
```
@cnpetra should we remove this line too?

cnpetra replied:
please

nychiang commented on src/Drivers/IpoptAdapter.hpp:
```diff
@@ -667,5 +667,406 @@ class hiopMDS2IpoptTNLP : public TNLP
+      if(!bret) return false;
+      nnzit += m_nnz_sparse_Hess_Lagr;
+
+#ifdef DEBUG
```
There are 10-15 places use `ifdef DEBUG`, e.g., in [hiopVectorRajaPar ](https://github.com/LLNL/hiop/blob/a1f62c21a1d3640b124c2af614567d6d978b7961/src/LinAlg/hiopVectorRajaPar.cpp#L450)and [hiopVectorPar](https://github.com/LLNL/hiop/blob/a1f62c21a1d3640b124c2af614567d6d978b7961/src/LinAlg/hiopVectorPar.cpp#L208) .

Maybe we should clarify the difference and file an issue.

cnpetra replied:
DEEPCHECKS is like DEBUG but should be used for expensive tests/asserts, such as inside loops. the idea is to be able to disable these expensive checks even in DEBUG

nychiang commented on src/LinAlg/hiopMatrixSparseTriplet.cpp:
```diff
@@ -594,4 +734,443 @@ startingAtAddSubDiagonalToStartingAt(int diag_src_start, const double& alpha,
+
+
+// Generate the three vectors A, IA, JA
+void hiopMatrixSparseTriplet::convertToCSR(int &csr_nnz, int **csr_kRowPtr_in, int **csr_jCol_in, double **csr_kVal_in,
```
this function needs to be rewritten, since I realize that std::unordered_map is not available on GPU.

nychiang commented on src/LinAlg/hiopMatrixSparseTriplet.cpp:
```diff
@@ -594,4 +734,443 @@ startingAtAddSubDiagonalToStartingAt(int diag_src_start, const double& alpha,
 } //end of namespace
+
+
+#if 0
```
let's remove this. This is another way to implement adding sparse rows into a big matrix.

cnpetra commented on CMakeLists.txt:
```diff
@@ -159,21 +163,29 @@ if(HIOP_WITH_KRON_REDUCTION)
+endif(NOT DEFINED BLAS_LIBRARIES)
+
+
+#set(LAPACK_LIBRARIES "-lgfortran;/usr/lib64/liblapack.so;-lpthread;-lm;-ldl")
```
yes, please remove.

pelesh commented:
In file `hiopNlpFormulation.cpp` there is quite a bit of indentation mess, probably due to text editor going berserk. This may obscure substantive changes to the code here. @nychiang is there an easy way to address this?

nychiang commented:
This is due to the extra trailing blanks. Have tuned my editor to ignore these issue, but it seems that I forgot to modify this file (and probably some others). The easiest way for me to correct this is run `diff -w` and then copy my inputs to the original file. 
just corrected this file in commit 8050d12

pelesh commented on tests/LinAlg/matrixTestsSparse.hpp:
```diff
@@ -732,6 +732,74 @@ class MatrixTestsSparse : public TestBase
+    B.setToConstant(B_val);
+
+    auto itnz_src{0};
+    auto itnz_dest=B_nnz_st;
```
This variable is not used and triggers compiler warnings. Can we remove it?

pelesh commented on tests/LinAlg/matrixTestsRajaSparseTriplet.hpp:
```diff
@@ -85,10 +85,12 @@ class MatrixTestsRajaSparseTriplet : public MatrixTestsSparse
   virtual const local_ordinal_type* getColumnIndices(const hiop::hiopMatrixSparse* a);
   virtual local_ordinal_type getLocalSize(const hiop::hiopVector *x) override;
   virtual int verifyAnswer(hiop::hiopMatrixSparse* A, real_type answer) override;
+  virtual int verifyAnswer(hiop::hiopMatrix* A, local_ordinal_type nnz_st, local_ordinal_type nnz_ed, const double answer) {assert(false);};
```
Provide return value to suppress compiler warnings:
```suggestion
  virtual int verifyAnswer(hiop::hiopMatrix* A, local_ordinal_type nnz_st, local_ordinal_type nnz_ed, const double answer) {assert(false); return 0;}
```

pelesh reviewed (approved):
There are quite a few code formatting issues, but other than that, the code seems to be working correctly. I will create another issue about that. It would be good to address as many of these as possible in this PR.

I also pointed out to some compiler warnings and misleading function names. None of these should block merging this PR, but it would be good to address as many of those as possible.

nychiang commented on cmake/FindHiopSTRUMPACK.cmake:
```diff
@@ -0,0 +1,42 @@
+    PATHS ${STRUMPACK_DIR} ${HIOP_STRUMPACK_DIR}
+    REQUIRED)
+    
+find_library(SCALAPACK_LIBRARY
```
I think the current version is just a temporary solution. Since Strumpack may be compiled with many other packages, we need to think about how to detect all its dependencies.
In my opinion, the ideal way is that the strumpack-config.cmake (generated by strumpack) should point out if ScaLapack is required and where to find it. HiOp can remove the dependency of scalapack if Strumpack can find it, as it can find and include metis, par-metis and other dependencies.

nychiang commented on src/Interface/README.md:
```diff
@@ -1,6 +1,7 @@
-HiOp supports two input formats: `hiopInterfaceDenseConstraints` and `hiopInterfaceMDS`. Both formats are in the form of C++ interfaces (e.g., abstract classes), see hiopInterface.hpp file, that the user must instantiate/implement and provide to HiOp.
+HiOp supports two input formats: `hiopInterfaceDenseConstraints`,`hiopInterfaceSparse` and `hiopInterfaceMDS`.
+Both formats are in the form of C++ interfaces (e.g., abstract classes), see [hiopInterface.hpp](hiopInterface.hpp) file, that the user must instantiate/implement and provide to HiOp.
```
:-) this has been addressed in branch `dev-nlpsparse-doc`. We are still working on polishing the documents in branches `dev-nlpsparse-doc` and `nlpsparse-doc`. Will file another PR when they are ready.

nychiang commented on src/LinAlg/hiopLinSolverSparseSTRUMPACK.hpp:
```diff
@@ -0,0 +1,187 @@
+  int     *jCol_;                      // column indexes for nonzeros
+  double  *kVal_;                      // storage for sparse matrix
+
+  int *index_covert_CSR2Triplet_;
+  int *index_covert_extra_Diag2CSR_;
```
Will do this after this PR. These two arrays are used when transferring a triplet form to CSR form, and they are only used in function `convertToCSR`. Our plan is that to rewrite some parts of that function after this PR. Currently I use std::unordered_map in it, but I learned that std containers are not available on GPU.

nychiang commented:
@pelesh Except for the ones where I left comments, I have addressed all the other issues.
Also noticed a memory leak when using strumpack. Fixed as well.

cnpetra commented:
> There are quite a few code formatting issues, but other than that, the code seems to be working correctly. I will create another issue about that. It would be good to address as many of these as possible in this PR.
> 
> I also pointed out to some compiler warnings and misleading function names. None of these should block merging this PR, but it would be good to address as many of those as possible.

please fill issues or at least dump your notes so I and NY are aware of them.

ashermancinelli reviewed (approved):
I echo slaven's comments, I think this is now ready for merge.
